src/librustc_mir/transform/fragment_locals.rs

```diff
@@ -0,0 +1,499 @@
+
+        (self, proj_elems)
+    }
+
+    fn assign_locals(
+        &mut self,
+        local_decls: &mut IndexVec<Local, LocalDecl<'tcx>>,
+        source_info: SourceInfo,
+    ) {
+        match self.kind {
+            FragmentTreeNodeKind::OpaqueLeaf { ref mut replacement_local } => {
+                let mut decl = LocalDecl::new_internal(self.ty, source_info.span);
+                decl.source_info = source_info;
+                *replacement_local = Some(local_decls.push(decl));
+            }
+            FragmentTreeNodeKind::Nested(ref mut fragments) => {
+                for fragment in fragments.values_mut() {
+                    fragment.assign_locals(local_decls, source_info);
+                }
+            }
+        }
+    }
+
+    /// Push debuginfo for all leaves into `fragments`, pointing to
+    /// their respective `replacement_local`s (set by `assign_locals`).
+    fn gather_debug_info_fragments(
+        &self,
+        dbg_fragment_projection: &mut Vec<ProjectionKind>,
+        dbg_fragments: &mut Vec<VarDebugInfoFragment<'tcx>>,
+    ) {
+        match self.kind {
+            FragmentTreeNodeKind::OpaqueLeaf { replacement_local } => {
+                dbg_fragments.push(VarDebugInfoFragment {
+                    projection: dbg_fragment_projection.clone(),
+                    contents: Place::from(replacement_local.expect("missing replacement")),
+                })
+            }
+            FragmentTreeNodeKind::Nested(ref fragments) => {
+                for (&f, fragment) in fragments {
+                    match f {
+                        Fragment::Discriminant => unreachable!(),
+                        Fragment::Field(variant_index, field) => {
+                            assert_eq!(variant_index, None);
+
+                            dbg_fragment_projection.push(ProjectionElem::Field(field, ()));
+                        }
+                    }
+                    fragment.gather_debug_info_fragments(dbg_fragment_projection, dbg_fragments);
+                    dbg_fragment_projection.pop();
+                }
+            }
+        }
+    }
+}
+
+struct FragmentTreeCollector<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    locals: IndexVec<Local, FragmentTree<'tcx>>,
+}
+
+impl FragmentTreeCollector<'tcx> {
+    fn place_node(&'a mut self, place: &Place<'tcx>) -> Option<&'a mut FragmentTree<'tcx>> {
+        let (node, proj_elems) = self.locals[place.local].project(place.projection);
+        if proj_elems.is_empty() { Some(node) } else { None }
+    }
+}
+
+impl Visitor<'tcx> for FragmentTreeCollector<'tcx> {
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, _: Location) {
+        // Borrows of fields might be used to access the entire local,
+        // by unsafe code, so it's better for the time being to remain
+        // conservative, until such uses have been definitely deemed UB.
+        if context.is_borrow() {
+            self.locals[place.local].make_opaque();
+        }
```

Comment:
Note to self: this is not catching `Rvalue::AddressOf` (which has a different `context`).

This could've been caught by tests, if we had `&raw` versions of `ui/mir/mir_raw_fat_ptr.rs` and `ui/raw-fat-ptr.rs` (which are the tests that failed before I added this special-case).
I doubt I'll do that (adding tests) in this PR, but maybe other people are interested.
cc @matthewjasper @RalfJung

Also I'm worried we might have code elsewhere that isn't handling `&raw` correctly, if it relies on either `Borrow` contexts (by pattern-matching or `is_borrow`) or `Rvalue::Ref`.
cc @ecstatic-morse @oli-obk

Reply:
I verified that `Rvalue::AddressOf` was handled correctly in const checking via explicit match in the PR that introduced `&raw`. It seems clear to me that `AddressOf` should not be its own `PlaceContext`, but should instead map to `SharedBorrow` and `Borrow`. I can't say whether other passes in the compiler are matching on `AddressOf` correctly.

Reply:
I wouldn't mind there being something like `BorrowOrAddrOf` and `is_borrow_or_addr_of` if we want to be super explicit, but otherwise I think I agree.

Reply:
>  if we had &raw versions of ui/mir/mir_raw_fat_ptr.rs and ui/raw-fat-ptr.rs

Those two tests look almost identical (the former just has some more stuff)... any idea why we have that duplication? One was clearly created by copy-pasting the other, but there are no comments, unfortunately.

Reply:
If it has mir in the name it used to have an opt-in to force mir codegen back when we had two modes (and `-Zorbit`). I agree we should dedup, cc @Centril